rewrite-java/src/main/java/org/openrewrite/java/tree/J.java

```diff
@@ -317,33 +317,82 @@ public String toString() {
+    @RequiredArgsConstructor
+    @With
     @Data
     final class ArrayType implements J, TypeTree, Expression {
-        @With
         @EqualsAndHashCode.Include
         UUID id;
 
-        @With
         Space prefix;
-
-        @With
         Markers markers;
-
-        @With
         TypeTree elementType;
 
-        @With
-        List<JRightPadded<Space>> dimensions;
+        @Nullable
+        List<J.Annotation> annotations;
 
-        @Override
-        public JavaType getType() {
-            return elementType.getType();
-        }
+        JLeftPadded<Space> dimension;
 
-        @SuppressWarnings("unchecked")
-        @Override
-        public ArrayType withType(@Nullable JavaType type) {
-            return type == getType() ? this : withElementType(elementType.withType(type));
+        JavaType type;
+
+        /**
+         * For backwards compatibility with older LSTs.
+         * Do not remove until we're confident older LSTs are no longer in use.
+         */
+        @Deprecated
+        @JsonCreator
+        ArrayType(UUID id,
+                  Space prefix,
+                  Markers markers,
+                  TypeTree elementType,
+                  @Nullable List<JRightPadded<Space>> dimensions, // Do not remove or rename, required for backwards compatibility.
+                  @Nullable List<J.Annotation> annotations,
+                  @Nullable JLeftPadded<Space> dimension,
+                  @Nullable JavaType type) {
+            this.id = id;
+            this.prefix = prefix;
+            this.markers = markers;
+            if (dimensions != null) {
+                JavaType javaType = elementType.getType() == null ? JavaType.Unknown.getInstance() : elementType.getType();
+                TypeTree base = dimensions.isEmpty() || dimensions.size() == 1 ? elementType : new ArrayType(
+                        Tree.randomId(),
+                        Space.EMPTY,
+                        Markers.EMPTY,
+                        elementType,
+                        null,
+                        dimensions.isEmpty() ? JLeftPadded.build(Space.EMPTY) :
+                                JLeftPadded.build(dimensions.get(0).getAfter()).withBefore(dimensions.get(0).getElement()),
+                        javaType
+                );
+                for (int i = 1; i < dimensions.size() - 1; i++) {
+                    JRightPadded<Space> dim = dimensions.get(i);
+                    base = new ArrayType(
+                            Tree.randomId(),
+                            Space.EMPTY,
+                            Markers.EMPTY,
+                            base,
+                            null,
+                            JLeftPadded.build(dim.getAfter()).withBefore(dim.getElement()),
+                            javaType
+                    );
+                }
+                this.elementType = base;
+                this.type = javaType;
```

Comment:
Possibly we can create a test case for this branch where we create an `ArrayType` using the old properties. I get the impression that type `type` field isn't quite correctly initialized yet. It looks like it will be equal to the element type's type attribution. Also the element type's type attribution looks wrong in this respect (see line 376 where `javaType` is reused).

Reply:
I'm guessing that you expect the type on the element is a `JavaType.Array`?

I had checked the types based on the previous code and found the element type was a `JavaType$Class` instead.

I.E., for the code `String[][]` the type was `java.lang.String`. The difference occurred because of code like:

```java
            JCExpression elementType = ((JCArrayTypeTree) vartype).elemtype;
            while (elementType instanceof JCArrayTypeTree) {
                elementType = ((JCArrayTypeTree) elementType).elemtype;
            }
```
At least for the Java parsers, the type on the element AFTER it is not an ArrayTypeTree is the class type, which does not happen anymore. So, the tests that assert the types are array types with dimensions is new.

However, the GroovyParser does set the J.Identifier to a `JavaType$Array` of `String[][]`. So, I can revise the code to detect a `JavaType$Array` on the `elementType` and make the appropriate transformation.

Reply:
Yes, for any `J.ArrayType` I think `getType()` should return a `JavaType.Array` instance. Since it didn't do that before AFAICT, I guess we could get away with not correcting that for old LSTs, but it isn't quite clear to me why we shouldn't be able to address that here.

Reply:
I've updated the constructor to create the new `J.Array` with the correct types, and added a test.